video_core: Address various UE bugs

raphaelthegreat opened the pull request.

raphaelthegreat committed:
vk_rasterizer: Reorder image query in fast clear elimination

Fixes missing clears when a texture is being cleared using this method but never actually used for rendering purposes by ensuring the texture cache has at least a chance to register cmask

raphaelthegreat committed:
shader_recompiler: Partial support for ANCILLARY_ENA

raphaelthegreat committed:
pixel_format: Add number conversion of BC6 srgb format

raphaelthegreat committed:
texture_cache: Support aliases of 3D and 2D array images

Used be UE to render its post processing LUT

squidbus commented on src/shader_recompiler/backend/spirv/spirv_emit_context.cpp:
```diff
@@ -370,13 +370,14 @@ void EmitContext::DefineInputs() {
         }
+        if (info.loads.GetAny(IR::Attribute::RenderTargetIndex)) {
+            output_layer = DefineVariable(S32[1], spv::BuiltIn::Layer, spv::StorageClass::Input);
+            Decorate(output_layer, spv::Decoration::Flat);
```
Should we rename to just `layer` or something since its not longer just used for output?

raphaelthegreat replied:
Technically output could be interpreted as the output render target, which makes sense in either stage which is why I didn't rename it. But I can do this if you prefer

squidbus replied:
Sure, should probably at least rename the info flag to `has_output_layer` to make more sense then.

squidbus commented on src/shader_recompiler/backend/spirv/emit_spirv.cpp:
```diff
@@ -305,6 +305,9 @@ void SetupCapabilities(const Info& info, const Profile& profile, const RuntimeIn
             ctx.AddCapability(spv::Capability::SampleRateShading);
         }
+        if (info.has_layer_output) {
+            ctx.AddCapability(spv::Capability::Geometry);
```
I'm not quite sure I understand this, fragment shaders can use the Geometry execution model?

raphaelthegreat replied:
Yeah I compiled a simple example with glslang and it added Geometry capability. Validation layers also complain if its missing
```
     OpCapability Geometry
%1 = OpExtInstImport "GLSL.std.450"
     OpMemoryModel Logical GLSL450
     OpEntryPoint Fragment %main "main" %color %gl_Layer
```

squidbus replied:
That's a weird case. The Vulkan spec side says that `Fragment` execution model should be fine (https://registry.khronos.org/vulkan/specs/latest/man/html/Layer.html), but if validation is complaining, glslang does it, and this works, then no reason not to I guess.

raphaelthegreat replied:
Here is the exact error for future reference
```
[Render.Vulkan] <Info> vk_pipeline_cache.cpp:502 CompileModule: Compiling fs shader 0xbf93e1b9a9b13256 
[Render.Vulkan] <Error> vk_platform.cpp:54 DebugUtilsCallback: VUID-VkShaderModuleCreateInfo-pCode-08737: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Operand 3 of Decorate requires one of these capabilities: Geometry ShaderLayer ShaderViewportIndexLayerEXT MeshShadingNV MeshShadingEXT 
  OpDecorate %gl_Layer BuiltIn Layer
The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must adhere to the validation rules described by the Validation Rules within a Module section of the SPIR-V Environment appendix (https://docs.vulkan.org/spec/latest/chapters/shaders.html#VUID-VkShaderModuleCreateInfo-pCode-08737)
```

squidbus commented on src/shader_recompiler/ir/passes/shader_info_collection_pass.cpp:
```diff
@@ -160,7 +160,8 @@ void CollectShaderInfoPass(IR::Program& program, const Profile& profile) {
-    if (info.stores.GetAny(IR::Attribute::RenderTargetIndex)) {
+    if (info.stores.GetAny(IR::Attribute::RenderTargetIndex) ||
+        info.loads.GetAny(IR::Attribute::RenderTargetIndex)) {
         info.has_layer_output = true;
```
Same for this, should probably rename since its also input now.

squidbus commented on src/video_core/amdgpu/pixel_format.h:
```diff
@@ -102,6 +102,7 @@ enum class NumberConversion : u32 {
     Sint8ToSnormNz = 4,
     Sint16ToSnormNz = 5,
     Uint32ToUnorm = 6,
+    SrgbToNorm = 7,
```
The order in these conversion op names is usually in the direction from the base format to what the shader sees, so I think this would be `SnormToSrgb` (or just `NormToSrgb` if the same code works for both sign types).

raphaelthegreat replied:
Isn't this the case already however, since srgb is the stored format and the shader will "see" linear norm values

squidbus replied:
It seems a bit confusing in this case. Normally for example the Uscaled conversion is a `Uint` Vulkan format converted to a `Uscaled` value in the shader, so it is `UintToUscaled`, and here we use `Snorm` on the Vulkan side. But I suppose you're right that in the end we're degammaing it and giving it to the shader as a linear value, so I guess it's fine?

squidbus commented on src/video_core/texture_cache/texture_cache.cpp:
```diff
@@ -285,6 +285,12 @@ std::tuple<ImageId, int, int> TextureCache::ResolveOverlap(const ImageInfo& imag
     if (image_info.guest_address == tex_cache_image.info.guest_address) { // Equal address
+        if (image_info.guest_size == tex_cache_image.info.guest_size &&
+            (image_info.type == AmdGpu::ImageType::Color3D ||
+             tex_cache_image.info.type == AmdGpu::ImageType::Color3D)) {
```
Noticed the commit message for this change, if we're doing this for 2D array / 3D aliases shouldn't we check that one of them is 2D / 2D array?

There's `IsViewTypeCompatible` in `image_view.cpp` if you want to reuse some of that as well but for image types.

raphaelthegreat replied:
Yeah I'm still not too happy with this, needs some extra checks and properly be merged into some other cases bellow. That path in general needs some cleanup (i.e tiling check could be removed in favor of checking array_mode when searching overlaps). Also might be better to split this into cases of >, =, < guest size and do actions depending on each one (copy, or make image view). Some of the code here is remnant from when this function only checked the size and not the more proper block dim/block size/tile mode checks that ensure proper compatibility

raphaelthegreat committed:
pixel_format: Test BC6 srgb as unorm

Still not sure what is up with snorm/unorm can be useful to have both actions to compare for now

raphaelthegreat committed:
video_core: Use attachment feedback layout instead of general if possible

UE games often do mipgen passes where the previous mip of the image being rendered to is bound for reading. This appears to cause corruption issues so use attachment feedback loop extension to ensure correct output

squidbus commented on src/video_core/renderer_vulkan/vk_rasterizer.cpp:
```diff
@@ -755,7 +754,9 @@ void Rasterizer::BindTextures(const Shader::Info& stage, Shader::Backend::Bindin
             if (image.binding.force_general || image.binding.is_target) {
-                image.Transit(vk::ImageLayout::eGeneral,
+                image.Transit(instance.IsAttachmentFeedbackLoopLayoutSupported()
+                                  ? vk::ImageLayout::eAttachmentFeedbackLoopOptimalEXT
```
You need to set `VK_IMAGE_USAGE_ATTACHMENT_FEEDBACK_LOOP_BIT_EXT` usage on the image to use this.

raphaelthegreat replied:
Looks like that also needs a pipeline compile flag too hmm. Should I also make VK_EXT_attachment_feedback_loop_dynamic_state mandatory to use feedback loops at all

squidbus replied:
```
Feedback loop is enabled for an attachment if either of the following conditions are satisfied:
* The corresponding image is in the VK_IMAGE_LAYOUT_ATTACHMENT_FEEDBACK_LOOP_OPTIMAL_EXT layout
* For the attachment identified by VkRenderingAttachmentInfo::feedbackLoopEnable is VK_TRUE
```

~~Seems to me like it's an either/or from the spec?~~

I think I'm looking at the wrong thing but the pipeline flag doesn't seem to be mentioned in the feedback loop section.

raphaelthegreat replied:
So probably need both pipeline flag and image usage flag seems like. The flag I'm referring to is VK_PIPELINE_CREATE_COLOR_ATTACHMENT_FEEDBACK_LOOP_BIT_EXT

squidbus replied:
Probably should be dynamic then yes

raphaelthegreat replied:
Newest commit should be a lot better. I'm a bit worried about performance impact of setting that usage flag on many color images, it should be checked ig

raphaelthegreat committed:
renderer_vulkan: Improve feedback loop code

* Set proper usage flag for feedback loop usage
* Add dynamic state extension and enable it for color aspect when necessary
* Check if image is bound instead of force_general for better code consistency

raphaelthegreat committed:
shader_recompiler: More proper depth export implementation

raphaelthegreat committed:
shader_recompiler: Fix bug in output modifiers

raphaelthegreat committed:
shader_recompiler: Fix sampling from MSAA images

This is not allowed by any graphics API but seems hardware supports it somehow and it can be encountered. To avoid glitched output translate to to a texelFetch call on sample 0

raphaelthegreat committed:
clang format

raphaelthegreat committed:
image: Add back missing code

raphaelthegreat committed:
shader_recompiler: Better ancillary implementation

Now is implemented with a custom attribute that is constant propagated depending on which parts of it are extracted. It will assert if an unknown part is used or if the attribute itself is not removed by dead code elim

raphaelthegreat committed:
copy_shader: Ignore not enabled export channels

raphaelthegreat committed:
constant_propagation: Invalidate ancillary after successful elimination

raphaelthegreat committed:
spirv: Fix f11/f10 conversion to f32

georgemoralis committed:
Merge branch 'main' into ue-fixes
